smolyay/basis.py

```diff
@@ -180,6 +180,145 @@ def make_nested_set(cls, exactness):
+
+        for :math:`n \ge 2`, and by the direct formula for the other values of *n*.
+
+        Parameters
+        ----------
+        x : float
+            One-dimensional point on :math:`[-1,1]`.
+
+        Returns
+        -------
+        float
+            Value of Chebyshev polynomial of the second kind.
+
+        Raises
+        -------
+        float
+            Value of Chebyshev polynomial of the second kind.
+
+        Raises
+        ------
+        ValueError
+            if input is outside the domain [-1,1]
+
+        """
+        if x > 1 or x < -1:
+            raise ValueError("Input is outside the domain [-1,1]")
+        if self._n == 0:
+            return 1
+        elif self._n == 1:
+            return 2*x
+        else:
+            k_lim = self._n//2
+            answer = 0
+            for k in range(0,k_lim+1):
+                answer += math.comb(self._n+1,2*k+1)*((x**2 - 1)**k)*(x**(self._n-2*k))
+            return answer
+
+    @classmethod
+    def make_nested_set(cls, exactness):
+        """Create a nested set of Chebyshev polynomials.
+
+        A nested set is created up to a given level of ``exactness``,
+        which corresponds to a highest-order Chebyshev polynomial of
+        degree ``n = 2**(exactness + 1) - 1``.
+
+        Each nesting level corresponds to the increasing powers of 2 going up to
+        ``2**(exactness + 1) - 1``, with the first level being a special case. 
+        The generating Chebyshev polynomials are hence of degree (1, 3, 7, 
+        15, ...). Each new point added in a level is paired with a basis 
+        function of increasing order.
+
+        For example, for an ``exactness`` of 3, the generating polynomials are
+        of degree 1, 3, 7, and 16, at each of 4 levels. There are 2, 2, 4, and 8
+        new points added at each level. The polynomials at level 0 are of degree
+        0 and 1, the polynomials at level 1 are of degrees 2 and 3, those at 
+        level 2 are of degree 4, 5, 6, and 7, and those at level 3 are of 
+        degrees 8, 9, 10, 11, 12, 13, 14, and 15.
+
+        Parameters
+        ----------
+        exactness : int
+            Level of exactness.
+
+        Returns
+        -------
+        NestedBasisFunctionSet
+            Nested Chebyshev polynomials of the first kind.
+
+        """
+        basis_functions = []
+        levels = []
+        points = []
+        for i in range(0, exactness+1):
+            if i > 0:
+                start_level = 2**i
+                end_level = 2**(i+1)-1
+            else:
+                start_level = 0
+                end_level = 1
+            level_range = range(start_level, end_level+1)
+
+            basis_functions.extend(ChebyshevSecondKind(n) for n in level_range)
+            levels.append(list(level_range))
+            for p in basis_functions[end_level].points:
+                if not numpy.isclose(points, p).any():
+                    points.append(p)
+        points = [0] + points
```

Comment:
Why do you prepend `[0]` here? Shouldn't it be generated already? This seems hacky.

Reply:
The issue is that the 0th level of the nested group contains 2 functions, the 0th degree and the 1st degree, while for the first kind the 0th level only has the 0th degree. Since it's set up to add points for the last degree in each level, it gets the 0 from 1 but not from 0.
I could replace the [0] on line 319 with ChebyshevSecondKind(0).points or with basis_functions[0].points
I could also do this:

`if i > 0:`
`    start_level = 2**i`
`    end_level = 2**(i+1)-1`
`else:`
`      start_level = 0`
`      end_level = 1`
`      points = ChebyshevSecondKind(0).points`

Reply:
OK, I think I see! Wouldn’t a better option be to handle the i=0 special case outside the loop, then start the loop counter from 1?

Reply:
So initialize with something like this?

`levels = [[0,1]]`
`basis_functions =  [ChebyshevSecondKind(n) for n in levels[0]]`
`points = basis_functions[0].points + basis_functions[1].points`

Reply:
Yes, although you could also just hard code the two basis functions (rather than do list comprehension) to be explicit. You should also add a comment explaining why you are doing this.

Reply:
Done